### ISSUE TITLE
Help with Teensy 3.5

### ISSUE THREAD

**Issue by mjs513** [felis/USB_Host_Shield_2.0#297]
Hi all.

I am using a homebrew shield for a teensy 3.5 breakout board that I made and am receiving the following error when using board_qc:
```
Circuits At Home 2011
USB Host Shield Quality Control Routine
Reading REVISION register... Die revision 03
SPI long test. Transfers 1MB of data. Each dot is 64K
Test failed.  Value written: 79 read: 40
Unrecoverable error - test halted!!
0x55 pattern is transmitted via SPI
Press RESET to restart test
```

Because it is a homebrew, I tested the shield separately on a Arduino 101 and it worked fine so I am pretty sure the shield is fine.  I tested the SPI interface on the breakout board using a Memory Card reader and the standard example for a teensy and had no problem with the tests.  I also tried using the official shield but it didn't even read the revision register.  I also checked the voltages to the board and they are ok.  I also see the light on the BT dongle.

Any suggestions would be appreciated, I am at a loss now.

Mike

**Comment by mjs513** [comment]
Just by way of an update I tried it with the version of UHS3.0 on my machine (probably out of date by now) but it worked fine.  Here is the output of the board_qc  test:
```
Circuits At Home 2011
USB Host Shield Quality Control Routine
Reading REVISION register... Die revision 03
SPI long test. Transfers 1MB of data. Each dot is 64K 
................
SPI long test passed
GPIO test. Connect GPIN0 to GPOUT7, GPIN1 to GPOUT6, and so on
Test failed. Wrote 0x00, read 0xff

Press any key to continue...
```

**Comment by xxxajk** [comment]
Totally normal, send a character to finish the test.

**Comment by mjs513** [comment]
@xxxajk Even though the revision level is invalid (thought it only went up to 3) ?

Mike

**Comment by xxxajk** [comment]
Rev3 is fine. Looks correct...

**Comment by mjs513** [comment]
Still can't get it to work with the 1.8.2 and teensyduino 1.36.  I did come across this post on the teensy forum about spi.  Not sure if it may be affecting SPI transfer or not but thought I should pass it along:
[https://forum.pjrc.com/threads/44413-SPI-transfer(bytes-count)-has-byte-order-issue-w-Teensy-3-2-and-Teensyduino-1-36](url)

which is mentioned it the new version of teensyduino coming out: [https://forum.pjrc.com/threads/44509-Arduino-1-8-3?p=144663&viewfull=1#post144663](url)

**Comment by mjs513** [comment]
Me Again.  Decided to do a little more debugging via modifying the board_qc.ino file to use the read/wrt routines directly from the INO which was copied from the UHS3.0.  Lo and behold it worked. Here is the results:

```
Circuits At Home 2011
USB Host Shield Quality Control Routine
Reading REVISION register... Die revision 03
SPI long test. Transfers 1MB of data. Each dot is 64K................ SPI long test passed
GPIO test. Connect GPIN0 to GPOUT7, GPIN1 to GPOUT6, and so on
Test failed. Value written: 00 Value read: FF 
Press any key to continue...
GPIO test passed.
PLL test. 100 chip resets will be performed
Resetting oscillator

Reset number 0 Time to stabilize - 1423 cycles
Reset number 1 Time to stabilize - 1421 cycles
Reset number 2 Time to stabilize - 1422 cycles
Reset number 3 Time to stabilize - 1424 cycles
Reset number 4 Time to stabilize - 1422 cycles
Reset number 5 Time to stabilize - 1422 cycles
Reset number 6 Time to stabilize - 1421 cycles
Reset number 7 Time to stabilize - 1423 cycles
Reset number 8 Time to stabilize - 1424 cycles
Reset number 9 Time to stabilize - 1422 cycles
Reset number 10 Time to stabilize - 1422 cycles
Reset number 11 Time to stabilize - 1422 cycles
Reset number 12 Time to stabilize - 1423 cycles
Reset number 13 Time to stabilize - 1423 cycles
Reset number 14 Time to stabilize - 1422 cycles
Reset number 15 Time to stabilize - 1422 cycles
Reset number 16 Time to stabilize - 1423 cycles
Reset number 17 Time to stabilize - 1423 cycles
Reset number 18 Time to stabilize - 1422 cycles
Reset number 19 Time to stabilize - 1422 cycles
Reset number 20 Time to stabilize - 1422 cycles
Reset number 21 Time to stabilize - 1423 cycles
Reset number 22 Time to stabilize - 1422 cycles
Reset number 23 Time to stabilize - 1422 cycles
Reset number 24 Time to stabilize - 1422 cycles
Reset number 25 Time to stabilize - 1422 cycles
Reset number 26 Time to stabilize - 1423 cycles
Reset number 27 Time to stabilize - 1422 cycles
Reset number 28 Time to stabilize - 1422 cycles
Reset number 29 Time to stabilize - 1422 cycles
Reset number 30 Time to stabilize - 1423 cycles
Reset number 31 Time to stabilize - 1422 cycles
Reset number 32 Time to stabilize - 1422 cycles
Reset number 33 Time to stabilize - 1421 cycles
Reset number 34 Time to stabilize - 1422 cycles
Reset number 35 Time to stabilize - 1423 cycles
Reset number 36 Time to stabilize - 1422 cycles
Reset number 37 Time to stabilize - 1422 cycles
Reset number 38 Time to stabilize - 1422 cycles
Reset number 39 Time to stabilize - 1423 cycles
Reset number 40 Time to stabilize - 1422 cycles
Reset number 41 Time to stabilize - 1422 cycles
Reset number 42 Time to stabilize - 1422 cycles
Reset number 43 Time to stabilize - 1422 cycles
Reset number 44 Time to stabilize - 1423 cycles
Reset number 45 Time to stabilize - 1423 cycles
Reset number 46 Time to stabilize - 1422 cycles
Reset number 47 Time to stabilize - 1422 cycles
Reset number 48 Time to stabilize - 1423 cycles
Reset number 49 Time to stabilize - 1422 cycles
Reset number 50 Time to stabilize - 1422 cycles
Reset number 51 Time to stabilize - 1422 cycles
Reset number 52 Time to stabilize - 1422 cycles
Reset number 53 Time to stabilize - 1423 cycles
Reset number 54 Time to stabilize - 1422 cycles
Reset number 55 Time to stabilize - 1422 cycles
Reset number 56 Time to stabilize - 1422 cycles
Reset number 57 Time to stabilize - 1422 cycles
Reset number 58 Time to stabilize - 1423 cycles
Reset number 59 Time to stabilize - 1422 cycles
Reset number 60 Time to stabilize - 1422 cycles
Reset number 61 Time to stabilize - 1422 cycles
Reset number 62 Time to stabilize - 1424 cycles
Reset number 63 Time to stabilize - 1423 cycles
Reset number 64 Time to stabilize - 1422 cycles
Reset number 65 Time to stabilize - 1422 cycles
Reset number 66 Time to stabilize - 1422 cycles
Reset number 67 Time to stabilize - 1423 cycles
Reset number 68 Time to stabilize - 1422 cycles
Reset number 69 Time to stabilize - 1422 cycles
Reset number 70 Time to stabilize - 1422 cycles
Reset number 71 Time to stabilize - 1424 cycles
Reset number 72 Time to stabilize - 1422 cycles
Reset number 73 Time to stabilize - 1422 cycles
Reset number 74 Time to stabilize - 1422 cycles
Reset number 75 Time to stabilize - 1421 cycles
Reset number 76 Time to stabilize - 1421 cycles
Reset number 77 Time to stabilize - 1422 cycles
Reset number 78 Time to stabilize - 1422 cycles
Reset number 79 Time to stabilize - 1422 cycles
Reset number 80 Time to stabilize - 1423 cycles
Reset number 81 Time to stabilize - 1424 cycles
Reset number 82 Time to stabilize - 1422 cycles
Reset number 83 Time to stabilize - 1422 cycles
Reset number 84 Time to stabilize - 1422 cycles
Reset number 85 Time to stabilize - 1423 cycles
Reset number 86 Time to stabilize - 1422 cycles
Reset number 87 Time to stabilize - 1422 cycles
Reset number 88 Time to stabilize - 1422 cycles
Reset number 89 Time to stabilize - 1422 cycles
Reset number 90 Time to stabilize - 1424 cycles
Reset number 91 Time to stabilize - 1423 cycles
Reset number 92 Time to stabilize - 1422 cycles
Reset number 93 Time to stabilize - 1422 cycles
Reset number 94 Time to stabilize - 1423 cycles
Reset number 95 Time to stabilize - 1423 cycles
Reset number 96 Time to stabilize - 1422 cycles
Reset number 97 Time to stabilize - 1422 cycles
Reset number 98 Time to stabilize - 1422 cycles
Reset number 99 Time to stabilize - 1423 cycles

Checking USB device communication.

Reset complete. Waiting for the first SOF...
Getting device descriptor
Descriptor Length:	14
Descriptor type:	82
USB version:		14C3
Device class:		FB
Device Subclass:	9C
Device Protocol:	F1
Max.packet size:	86
Vendor  ID:		AE65
Product ID:		9305
Revision ID:		7E06
Mfg.string index:	BE
Prod.string index:	2F
Serial number index:	E2
Number of conf.:	B5

All tests passed. Press RESET to restart test
```

Sounds like something amiss with SPI.  Here is the sketchI used:
```
/* USB Host Shield 2.0 board quality control routine */
/* To see the output set your terminal speed to 115200 */
/* for GPIO test to pass you need to connect GPIN0 to GPOUT7, GPIN1 to GPOUT6, etc. */
/* otherwise press any key after getting GPIO error to complete the test */
/**/
#include <usbhub.h>

// Satisfy the IDE, which needs to see the include statment in the ino too.
#ifdef dobogusinclude
#include <spi4teensy3.h>
#include <../../../../hardware/pic32/libraries/SPI/SPI.h> // Hack to use the SPI library
#include <SPI.h> // Hack to use the SPI library
#endif

#include <SPI.h>

/* variables */
uint8_t rcode;
uint8_t usbstate;
uint8_t laststate;
//uint8_t buf[sizeof(USB_DEVICE_DESCRIPTOR)];
USB_DEVICE_DESCRIPTOR buf;

/* objects */
USB Usb;
//USBHub hub(&Usb);

void setup() {
        laststate = 0;
        Serial.begin(115200);
#if !defined(__MIPSEL__)
        while(!Serial); // Wait for serial port to connect - used on Leonardo, Teensy and other boards with built-in USB CDC serial connection
#endif
        E_Notify(PSTR("\r\nCircuits At Home 2011"), 0x80);
        E_Notify(PSTR("\r\nUSB Host Shield Quality Control Routine"), 0x80);
        /* SPI quick test - check revision register */
        E_Notify(PSTR("\r\nReading REVISION register... Die revision "), 0x80);
        Usb.Init(); // Initializes SPI, we don't care about the return value here
        {
                uint8_t tmpbyte = regRd(rREVISION);
                switch(tmpbyte) {
                        case( 0x01): //rev.01
                                E_Notify(PSTR("01"), 0x80);
                                break;
                        case( 0x12): //rev.02
                                E_Notify(PSTR("02"), 0x80);
                                break;
                        case( 0x13): //rev.03
                                E_Notify(PSTR("03"), 0x80);
                                break;
                        default:
                                E_Notify(PSTR("invalid. Value returned: "), 0x80);
                                print_hex(tmpbyte, 8);
                                halt55();
                                break;
                }//switch( tmpbyte...
        }//check revision register
        /* SPI long test */
        {
                E_Notify(PSTR("\r\nSPI long test. Transfers 1MB of data. Each dot is 64K"), 0x80);
                uint8_t sample_wr = 0;
                uint8_t sample_rd = 0;
                uint8_t gpinpol_copy = regRd(rGPINPOL);
                for(uint8_t i = 0; i < 16; i++) {
                        for(uint16_t j = 0; j < 65535; j++) {
                                regWr(rGPINPOL, sample_wr);
                                sample_rd = regRd(rGPINPOL);
                                if(sample_rd != sample_wr) {
                                        E_Notify(PSTR("\r\nTest failed.  "), 0x80);
                                        E_Notify(PSTR("Value written: "), 0x80);
                                        print_hex(sample_wr, 8);
                                        E_Notify(PSTR(" read: "), 0x80);
                                        print_hex(sample_rd, 8);
                                        halt55();
                                }//if( sample_rd != sample_wr..
                                sample_wr++;
                        }//for( uint16_t j...
                        E_Notify(PSTR("."), 0x80);
                }//for( uint8_t i...
                regWr(rGPINPOL, gpinpol_copy);
                E_Notify(PSTR(" SPI long test passed"), 0x80);
        }//SPI long test
        /* GPIO test */
        /* in order to simplify board layout, GPIN pins on text fixture are connected to GPOUT */
        /* in reverse order, i.e, GPIN0 is connected to GPOUT7, GPIN1 to GPOUT6, etc. */
        {
                uint8_t tmpbyte;
                E_Notify(PSTR("\r\nGPIO test. Connect GPIN0 to GPOUT7, GPIN1 to GPOUT6, and so on"), 0x80);
                for(uint8_t sample_gpio = 0; sample_gpio < 255; sample_gpio++) {
                        gpioWr(sample_gpio);
                        tmpbyte = Usb.gpioRd();
                        /* bit reversing code copied vetbatim from http://graphics.stanford.edu/~seander/bithacks.html#BitReverseObvious */
                        tmpbyte = ((tmpbyte * 0x0802LU & 0x22110LU) | (tmpbyte * 0x8020LU & 0x88440LU)) * 0x10101LU >> 16;
                        if(sample_gpio != tmpbyte) {
                                E_Notify(PSTR("\r\nTest failed. Value written: "), 0x80);
                                print_hex(sample_gpio, 8);
                                E_Notify(PSTR(" Value read: "), 0x80);
                                print_hex(tmpbyte, 8);
                                E_Notify(PSTR(" "), 0x80);
                                press_any_key();
                                break;
                        }//if( sample_gpio != tmpbyte...
                }//for( uint8_t sample_gpio...
                E_Notify(PSTR("\r\nGPIO test passed."), 0x80);
        }//GPIO test
        /* PLL test. Stops/starts MAX3421E oscillator several times */
        {
                E_Notify(PSTR("\r\nPLL test. 100 chip resets will be performed"), 0x80);
                /* check current state of the oscillator */
                if(!(regRd(rUSBIRQ) & bmOSCOKIRQ)) { //wrong state - should be on
                        E_Notify(PSTR("\r\nCurrent oscillator state unexpected."), 0x80);
                        press_any_key();
                }
                /* Restart oscillator */
                E_Notify(PSTR("\r\nResetting oscillator\r\n"), 0x80);
                for(uint16_t i = 0; i < 100; i++) {
                        E_Notify(PSTR("\rReset number "), 0x80);
                        Serial.print(i, DEC);
                        Usb.regWr(rUSBCTL, bmCHIPRES); //reset
                        if(regRd(rUSBIRQ) & bmOSCOKIRQ) { //wrong state - should be off
                                E_Notify(PSTR("\r\nCurrent oscillator state unexpected."), 0x80);
                                halt55();
                        }
                        regWr(rUSBCTL, 0x00); //release from reset
                        uint16_t j = 0;
                        for(j = 1; j < 65535; j++) { //tracking off to on time
                                if(regRd(rUSBIRQ) & bmOSCOKIRQ) {
                                        E_Notify(PSTR(" Time to stabilize - "), 0x80);
                                        Serial.print(j, DEC);
                                        E_Notify(PSTR(" cycles\r\n"), 0x80);
                                        break;
                                }
                        }//for( uint16_t j = 0; j < 65535; j++
                        if(j == 0) {
                                E_Notify(PSTR("PLL failed to stabilize"), 0x80);
                                press_any_key();
                        }
                }//for( uint8_t i = 0; i < 255; i++

        }//PLL test
        /* initializing USB stack */
        if(Usb.Init() == -1) {
                E_Notify(PSTR("\r\nOSCOKIRQ failed to assert"), 0x80);
                halt55();
        }
        E_Notify(PSTR("\r\nChecking USB device communication.\r\n"), 0x80);
}

void loop() {
        delay(200);
        Usb.Task();
        usbstate = Usb.getUsbTaskState();
        if(usbstate != laststate) {
                laststate = usbstate;
                /**/
                switch(usbstate) {
                        case( USB_DETACHED_SUBSTATE_WAIT_FOR_DEVICE):
                                E_Notify(PSTR("\r\nWaiting for device..."), 0x80);
                                break;
                        case( USB_ATTACHED_SUBSTATE_RESET_DEVICE):
                                E_Notify(PSTR("\r\nDevice connected. Resetting..."), 0x80);
                                break;
                        case( USB_ATTACHED_SUBSTATE_WAIT_SOF):
                                E_Notify(PSTR("\r\nReset complete. Waiting for the first SOF..."), 0x80);
                                break;
                        case( USB_ATTACHED_SUBSTATE_GET_DEVICE_DESCRIPTOR_SIZE):
                                E_Notify(PSTR("\r\nSOF generation started. Enumerating device..."), 0x80);
                                break;
                        case( USB_STATE_ADDRESSING):
                                E_Notify(PSTR("\r\nSetting device address..."), 0x80);
                                break;
                        case( USB_STATE_RUNNING):
                                E_Notify(PSTR("\r\nGetting device descriptor"), 0x80);
                                rcode = Usb.getDevDescr(1, 0, sizeof (USB_DEVICE_DESCRIPTOR), (uint8_t*) & buf);

                                if(rcode) {
                                        E_Notify(PSTR("\r\nError reading device descriptor. Error code "), 0x80);
                                        print_hex(rcode, 8);
                                } else {
                                        /**/
                                        E_Notify(PSTR("\r\nDescriptor Length:\t"), 0x80);
                                        print_hex(buf.bLength, 8);
                                        E_Notify(PSTR("\r\nDescriptor type:\t"), 0x80);
                                        print_hex(buf.bDescriptorType, 8);
                                        E_Notify(PSTR("\r\nUSB version:\t\t"), 0x80);
                                        print_hex(buf.bcdUSB, 16);
                                        E_Notify(PSTR("\r\nDevice class:\t\t"), 0x80);
                                        print_hex(buf.bDeviceClass, 8);
                                        E_Notify(PSTR("\r\nDevice Subclass:\t"), 0x80);
                                        print_hex(buf.bDeviceSubClass, 8);
                                        E_Notify(PSTR("\r\nDevice Protocol:\t"), 0x80);
                                        print_hex(buf.bDeviceProtocol, 8);
                                        E_Notify(PSTR("\r\nMax.packet size:\t"), 0x80);
                                        print_hex(buf.bMaxPacketSize0, 8);
                                        E_Notify(PSTR("\r\nVendor  ID:\t\t"), 0x80);
                                        print_hex(buf.idVendor, 16);
                                        E_Notify(PSTR("\r\nProduct ID:\t\t"), 0x80);
                                        print_hex(buf.idProduct, 16);
                                        E_Notify(PSTR("\r\nRevision ID:\t\t"), 0x80);
                                        print_hex(buf.bcdDevice, 16);
                                        E_Notify(PSTR("\r\nMfg.string index:\t"), 0x80);
                                        print_hex(buf.iManufacturer, 8);
                                        E_Notify(PSTR("\r\nProd.string index:\t"), 0x80);
                                        print_hex(buf.iProduct, 8);
                                        E_Notify(PSTR("\r\nSerial number index:\t"), 0x80);
                                        print_hex(buf.iSerialNumber, 8);
                                        E_Notify(PSTR("\r\nNumber of conf.:\t"), 0x80);
                                        print_hex(buf.bNumConfigurations, 8);
                                        /**/
                                        E_Notify(PSTR("\r\n\nAll tests passed. Press RESET to restart test"), 0x80);
                                        while(1);
                                }
                                break;
                        case( USB_STATE_ERROR):
                                E_Notify(PSTR("\r\nUSB state machine reached error state"), 0x80);
                                break;

                        default:
                                break;
                }//switch( usbstate...
        }
}//loop()...

/* constantly transmits 0x55 via SPI to aid probing */
void halt55() {

        E_Notify(PSTR("\r\nUnrecoverable error - test halted!!"), 0x80);
        E_Notify(PSTR("\r\n0x55 pattern is transmitted via SPI"), 0x80);
        E_Notify(PSTR("\r\nPress RESET to restart test"), 0x80);

        while(1) {
                Usb.regWr(0x55, 0x55);
        }
}

/* prints hex numbers with leading zeroes */
void print_hex(int v, int num_places) {
        int mask = 0, n, num_nibbles, digit;

        for(n = 1; n <= num_places; n++) {
                mask = (mask << 1) | 0x0001;
        }
        v = v & mask; // truncate v to specified number of places

        num_nibbles = num_places / 4;
        if((num_places % 4) != 0) {
                ++num_nibbles;
        }
        do {
                digit = ((v >> (num_nibbles - 1) * 4)) & 0x0f;
                Serial.print(digit, HEX);
        } while(--num_nibbles);
}

/* prints "Press any key" and returns when key is pressed */
void press_any_key() {
        E_Notify(PSTR("\r\nPress any key to continue..."), 0x80);
        while(Serial.available() <= 0); //wait for input
        Serial.read(); //empty input buffer
        return;
}

/* write single byte into MAX3421e register */
void regWr(uint8_t reg, uint8_t data) {
        SPI.beginTransaction(SPISettings(26000000, MSBFIRST, SPI_MODE0));
        digitalWriteFast(10, LOW);
        SPI.transfer(reg | 0x02);
        SPI.transfer(data);
        digitalWriteFast(10, HIGH);
        SPI.endTransaction();
}


/* multiple-byte write                            */

/* returns a pointer to memory position after last written */
uint8_t* bytesWr(uint8_t reg, uint8_t nbytes, uint8_t* data_p) {
        SPI.beginTransaction(SPISettings(26000000, MSBFIRST, SPI_MODE0));
        digitalWriteFast(10, LOW);
        SPI.transfer(reg | 0x02);
        //printf("%2.2x :", reg);

        while(nbytes) {
                SPI.transfer(*data_p);
                //printf("%2.2x ", *data_p);
                nbytes--;
                data_p++; // advance data pointer
        }
        digitalWriteFast(10, HIGH);
        SPI.endTransaction();
        //printf("\r\n");
        return (data_p);
}
/* GPIO write                                           */
/*GPIO byte is split between 2 registers, so two writes are needed to write one byte */

/* GPOUT bits are in the low nibble. 0-3 in IOPINS1, 4-7 in IOPINS2 */
void gpioWr(uint8_t data) {
        regWr(rIOPINS1, data);
        data >>= 4;
        regWr(rIOPINS2, data);
        return;
}

/* single host register read    */
uint8_t regRd(uint8_t reg) {
        SPI.beginTransaction(SPISettings(26000000, MSBFIRST, SPI_MODE0));
        digitalWriteFast(10, LOW);
        SPI.transfer(reg);
        uint8_t rv = SPI.transfer(0);
        digitalWriteFast(10, HIGH);
        SPI.endTransaction();
        return (rv);
}
/* multiple-byte register read  */

/* returns a pointer to a memory position after last read   */
uint8_t* bytesRd(uint8_t reg, uint8_t nbytes, uint8_t* data_p) {
        SPI.beginTransaction(SPISettings(26000000, MSBFIRST, SPI_MODE0));
        digitalWriteFast(10, LOW);
        SPI.transfer(reg);
        while(nbytes) {
                *data_p++ = SPI.transfer(0);
                nbytes--;
        }
        digitalWriteFast(10, HIGH);
        SPI.endTransaction();
        return ( data_p);
}

/* GPIO read. See gpioWr for explanation */

/* GPIN pins are in high nibbles of IOPINS1, IOPINS2    */
uint8_t gpioRd(void) {
        uint8_t gpin = 0;
        gpin = regRd(rIOPINS2); //pins 4-7
        gpin &= 0xf0; //clean lower nibble
        gpin |= (regRd(rIOPINS1) >> 4); //shift low bits and OR with upper from previous operation.
        return ( gpin);
}

/* reset MAX3421E. Returns number of microseconds it took for PLL to stabilize after reset
  or zero if PLL haven't stabilized in 65535 cycles */
uint16_t reset(void) {
        uint16_t i = 0;
        regWr(rUSBCTL, bmCHIPRES);
        regWr(rUSBCTL, 0x00);
        int32_t now;
        uint32_t expires = micros() + 65535;
        while((int32_t)(micros() - expires) < 0L) {
                if((regRd(rUSBIRQ) & bmOSCOKIRQ)) {
                        break;
                }
        }
        now = (int32_t)(micros() - expires);
        if(now < 0L) {
                i = 65535 + now; // Note this subtracts, as now is negative
        }
        return (i);
}
```

**Comment by xxxajk** [comment]
I think I know why...don't use spi4teensy3 and see if that solves the problem.
This can be changed by setting this macro to zero:
https://github.com/felis/USB_Host_Shield_2.0/blob/master/settings.h#L71
If that solves the problem, then it can be easily fixed in the selection code.

**Comment by mjs513** [comment]
Hmm.  That looked like it was already set to zero by: 
```
// Disabled on the Teensy LC, Teensy 3.5, and Teensy 3.6 as it is incompatible for now
#if defined(__MKL26Z64__) || defined(__MK64FX512__) || defined(__MK66FX1M0__)
#undef USE_SPI4TEENSY3
#define USE_SPI4TEENSY3 0
#endif
```
I gave it a try anyway this way:
```
#ifndef USE_SPI4TEENSY3
#define USE_SPI4TEENSY3 0
#endif
```
Any still did not work, here is the output:
Circuits At Home 2011
```
USB Host Shield Quality Control Routine
Reading REVISION register... Die revision 03
SPI long test. Transfers 1MB of data. Each dot is 64K
Test failed.  Value written: 62 read: 40
Unrecoverable error - test halted!!
0x55 pattern is transmitted via SPI
Press RESET to restart test
```

**Comment by xxxajk** [comment]
Ok, just making certain.

**Comment by mjs513** [comment]
Ok.  I updated the usbhost.h file with the read/wrt stuff from above and it worked.  In other words it read the revision, passed the SPI long writes, etc.  Everything was good.  I tried running the PS4USB sketch and just hung even with the uhs_debugging enabled.   So maybe now an interrupt issue.  Am wondering is anyone else is having this problem now? 

So the million dollar question is how to get this fixed?

**Comment by mjs513** [comment]
OK take my last post back.  Think I was getting crossed eyed.  Ran the usb sketch instead of the bt sketch.  THey both work now problem.

**Comment by mjs513** [comment]
Here are the files I modified - think someone needs to take a look and see what needs to be changed for possible incorporation.

[USB_Host_Shield_2.0-1.3.0.zip](https://github.com/felis/USB_Host_Shield_2.0/files/1049239/USB_Host_Shield_2.0-1.3.0.zip)

**Comment by mjs513** [comment]
OK.  Paul released teensyduino 1.37-beta1 today and it seems to have cleared up the problem.  No need to apply any changes to make it work now.  All is good.
